### PR TITLE
SAI bfd multihop changes along with state change notifications

### PR DESCRIPTION
**Pull request by abdbaig**
### Objective

Add BFD multihop support in SONIC SAI VPP.

NOTE: This will only work when BFD multihop changes are also present in VPP codebase.

### Changes
-Re-arrange existing BFD single-hop code
-Implement SAI changes to support BFD multihop session create and delete
-Add support to receive BFD notifications from VPP and propagate up to control plane so STATE_DB is updated with the latest BFD session state

### CODE REVIEW

File: platform/saivpp/vpplib/SwitchStateBaseRif.cpp [hunk 1]
```diff
@@ -44,6 +44,33 @@ using namespace saivpp;
 
 int SwitchStateBase::currentMaxInstance = 0;
 
+/* Utility function for IP addr translation from VPP to SAI */
+static void vpp_to_sai_ip_addr(
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseRif.cpp:48]: Let's move it to the central place SwitchStateUtils.h to be used by other modules

File: platform/saivpp/vpplib/SwitchStateBaseBfd.cpp [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+            SWSS_LOG_ERROR("BFD session create request FAILED due to invalid hwif name");
+    
+            return SAI_STATUS_FAILURE;
+        }
+    }
+
+    if (multihop || hwif_name) {
+        SWSS_LOG_NOTICE("BFD session create request sent to VPP, hwif: %s, multihop: %d, oid: %lld",  hwif_name, multihop, bfd_oid);
+        /*vpp call to add bfd session*/
+        ret = bfd_udp_add(multihop, hwif_name, &vpp_local_addr, &vpp_peer_addr, detect_mult, required_min_tx, required_min_rx);
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseBfd.cpp:147]: If you care about return, you need to change W(ret) in bfd_udp_add to WR(ret). W(ret) just returns the last result in the queue, which may not be the reply of your bfd request.

File: platform/saivpp/vpplib/SwitchStateBaseBfd.cpp [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+            SWSS_LOG_ERROR("BFD session delete request FAILED due to invalid hwif name");
+
+            return SAI_STATUS_FAILURE;
+        }
+    }
+
+    if (multihop || hwif_name) {
+        SWSS_LOG_NOTICE("BFD session delete request sent to VPP, hwif: %s, multihop: %d, oid: %lld",  hwif_name, multihop, bfd_oid);
+        /*vpp call to add bfd session*/
+        ret = bfd_udp_del(multihop, hwif_name, &vpp_local_addr, &vpp_peer_addr);
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseBfd.cpp:236]: Same here. Please change W(ret) in bfd_udp_del to WR(ret)

File: platform/saivpp/vpplib/SwitchStateBaseRif.cpp [hunk 4]
```diff
@@ -379,12 +406,66 @@ void SwitchStateBase::vppProcessEvents ()
...
+        break;
+    case VPP_API_BFD_STATE_UP:
+        sai_state = SAI_BFD_SESSION_STATE_UP;
+        break;
+    default:
+        sai_state = SAI_BFD_SESSION_STATE_DOWN;
+        break;
+    }
+
+    bfd_info.multihop = bfd_notif->multihop;
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseRif.cpp:447]: Do you need port to find single hop bfd session?

File: platform/saivpp/vpplib/SwitchStateBaseRif.cpp [hunk 5]
```diff
@@ -379,12 +406,66 @@ void SwitchStateBase::vppProcessEvents ()
...
+    vpp_to_sai_ip_addr(&bfd_info.local_addr, &bfd_notif->local_addr);
+    vpp_to_sai_ip_addr(&bfd_info.peer_addr, &bfd_notif->peer_addr);
+
+    // Find the BFD session
+    auto it = m_bfd_info_map.find(bfd_info);
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseRif.cpp:452]: std::map is not thread-safe and here we may have concurrent modification to the map from upstream. We need to protect the map using mutex.

File: platform/saivpp/vpplib/SwitchStateBaseBfd.cpp [hunk 6]
```diff
@@ -77,49 +53,83 @@ sai_status_t SwitchStateBase::vpp_bfd_session_add(
...
+    sai_ip_address_t peer_addr = attr->value.ipaddr;
+    vpp_ip_addr_t vpp_peer_addr;
+    sai_ip_address_t_to_vpp_ip_addr_t(peer_addr, vpp_peer_addr);
 
     /* Attribute#6 */
-    attr.id = SAI_BFD_SESSION_ATTR_MULTIHOP;
-    CHECK_STATUS(get(SAI_OBJECT_TYPE_BFD_SESSION, bfd_oid, 1, &attr));
-    bool multihop = attr.value.booldata;
+    attr = sai_metadata_get_attr_by_id(SAI_BFD_SESSION_ATTR_MULTIHOP, attr_count, attr_list);
+    if (!attr)
```
[the remark this answers is not in the document]
Comment (abdbaig) [reply platform/saivpp/vpplib/SwitchStateBaseBfd.cpp]: Will remove the return

**Review by yue-fred-gao — APPROVED** [review]

**Pull request merged**